### DISCUSSION TITLE
Introducing the .ng format in Analog [CLOSED]

### DISCUSSION THREAD

**Discussion by brandonroberts** [analogjs/analog#824] [Ideas]
This opens the discussion around https://github.com/analogjs/analog/pull/823. The goal is to get early feedback about the feature in its experimental phase.

This introduces **experimental** support for a `.ng` format for components, directives, and pipes. Initial work done by @nartc and  @brandonroberts, with early discussions also contributed by @joshuamorony.

## Why

- Forward-looking for new component authoring format
- Simpler mental model for component authoring
- Similar to other frameworks that's not JSX-based
- Doesn't change authoring services
- Works with existing concepts and components
- Works more seamlessly with new functions, lifecycle hooks 

## Notes

- This is enabled with an experimental flag through the Vite config.
- Components/directives can be used immediately in the template after imported.
- Uses Angular's `experimental-local` compilation mode.
- ~~SSR is not currently supported.~~
- ~~Inputs/Outputs are not currently supported.~~
- Deferrable Views are not currently supported.
- ~~`defineComponentMetadata` , `defineDirectiveMetadata`, and `definePipeMetadata` are used to define metadata for components, directives, and pipes.~~

#### 2024/01/02
- Support for custom pipes have been removed in favor of using `computed`
- `defineComponentMetadata` , `defineDirectiveMetadata` have been consolidated into `defineMetadata`. A component requires a `<template>` where a directive does not.

#### 2024/01/03
- Pipe metadata removed
- `exposes` property added to metadata for exposing globals, enums to the template
- `onDestroy` hook added to go along with `onInit`
- Object Destructure and Array Destructure are supported.
  - Destructuring reactive values will **NOT** work
  ```ts
  const data = signal({ firstName: 'chau', lastName: 'tran' });
  // if `data` updates, `firstName` and `lastName` will NOT update
  const { firstName, lastName } = data();
  ```

#### 2024/01/04

- `let` variables are supported by utilizing getters via `Object.defineProperties`
#### 2024/01/11

- Inputs are now supported via Angular 17.1 Signal Inputs
- Outputs are now supported via calling `new EventEmitter()` directly
  - No alias for Outputs for now.

#### 2024/01/25

- Support for .ng files is removed

### Concepts
```ts
/// <reference types="vitest" />

import { defineConfig } from 'vite';
import analog from '@analogjs/platform';

// https://vitejs.dev/config/
export default defineConfig(({ mode }) => ({
  // ... other config
  plugins: [
    analog({
      ssr: false,
      vite: {
        experimental: {
          dangerouslySupportNgFormat: true
        }
      },
    }),
  ],
}));
```

### Component

hello.ng
```html
<template>
  <p>hello works!</p>
</template>
```

app.component.ng
```html
<script lang="ts">
  import { inject, signal, effect, computed } from '@angular/core';
  import { JsonPipe } from '@angular/common';
  import { HttpClient } from '@angular/common/http';
  import { delay } from 'rxjs';

  import Hello from './hello.ng';
  import Highlight from './highlight.ng';
  import Doubled from './doubled.ng';

  defineMetadata({
    selector: 'app-root',
    imports: [JsonPipe],
  });

  const title = 'Analog';

  const http = inject(HttpClient);

  const counter = signal(1);
  const doubled = computed(() => counter() * 2);
  const todo = signal(null);

  const increment = () => {
    counter.update((value) => value + 1);
  };

  function decrement() {
    counter.update((value) => value - 1);
  }

  effect(() => {
    console.log('counter changed', counter());
  });

  onInit(() => {
    console.log('App init');
    http
      .get('https://jsonplaceholder.typicode.com/todos/1')
      .pipe(delay(2000))
      .subscribe((data) => {
        todo.set(data);
        console.log('data', data);
      });
  });
</script>

<template>
  @if (counter() > 2) {
    <Hello />
  }

  <p>Counter: {{ counter() }}</p>
  <p highlight>Doubled: {{ doubled() }}</p>
  <p>Doubled with Pipe: {{ counter() | doubled }}</p>
  <button (click)="increment()">increment</button>
  <button (click)="decrement()">decrement</button>

  @if (todo(); as todo) {
  <pre>{{todo | json }}</pre>
  } @else {
  <p>Loading todo...</p>
  }
</template>

<style>
  p {
    color: red;
  }

  button {
    background: blue;
    color: white;
    padding: 1rem 0.5rem;
    border-radius: 0.5rem;
  }
</style>
```

### Pipe (No longer included)

doubled.ng
```html
<script lang="ts">
  definePipeMetadata({
    name: 'doubled',
  });

  function transform(value: number) {
    return value * 2;
  }
</script>
```

### Directive

highlight.ng
```html
<script lang="ts">
  import { afterNextRender, ElementRef, inject } from '@angular/core';

  defineMetadata({
    selector: 'p[highlight]',
  });

  const elementRef = inject<ElementRef<HTMLElement>>(ElementRef);

  afterNextRender(() => {
    elementRef.nativeElement.style.backgroundColor = 'yellow';
  });
</script>
```

### Lifecycle Hooks

An `onInit` lifecycle hook can be added for init logic.

```html
<script>
  onInit(() => {
    console.log('App init');
    http
      .get('https://jsonplaceholder.typicode.com/todos/1')
      .pipe(delay(2000))
      .subscribe((data) => {
        console.log('data', data);
      });
  });
</script>

<template>
  <p>Component Init</p>
</template>
```

An `onDestroy` lifecycle hook can be added for teardown logic.

```html
<script lang="ts">
  onDestroy(() => {
    console.log('hello destroyed');
  });
</script>

<template>
  <p>Component Destroy</p>
</template>
```

Alternatively, `DestroyRef` can also be used for teardown logic instead:

```html
<script lang="ts">
  import { inject, DestroyRef } from '@angular/core';

  inject(DestroyRef).onDestroy(() => {
    console.log('hello destroyed');
  });
</script>
```

### Sample App

https://github.com/brandonroberts/analog-ng

**Comment by joshuamorony** [comment]
Pretty sure I don't need to reiterate how much I like/want this - thank you Brandon and Chau for spearheading this!

Just a couple of initial thoughts/nits:
* Not sure if it makes it easier to implement, but my preference would be for no `<template>` tag
* I think it feels somewhat unintuitive that adding init/destroy behaviour is handled differently - perhaps also add an `onDestroy` hook to keep this consistent?
* Perhaps something like `ngPipe()` and `ngDirective()` (or maybe just `defineDirective`) for the metadata instead of `defineDirectiveMetadata`- not sure what would be best but `defineDirectiveMetadata` feels a bit verbose

I also wonder whether the `.ng` file makes sense for directives/pipes? Perhaps rather than:

**doubled.ng**

We would stick with something like **doubled.pipe.ts** but instead of using decorators/classes it would be created entirely within the function, e.g:

**doubled.pipe.ts**
```ts
definePipe({
  name: 'doubled',
  transform: (value: number) => {
    return value * 2
  }
})
```

Of course, this is coming purely from the perspective of how I would ideally want this to work without considering how much this complicates the actual implementation

**Reply by nartc** [reply]
@joshuamorony 

1. `<template>` makes it easier to implement as well as create clear boundary for what goes in the component's template. I think I'd like to keep it around unless there are technical concerns as to why it shouldn't be there
2. `onDestroy` sounds good to me and it's not a big deal to add
3. In my personal opinion, Pipes aren't needed anymore since it's truly easier to create a pure function and pass it to `computed` or something. The only reason I support it is for "familiarity" purposes. And the main point of it being in a `.ng` file is Injection support and all that. It makes the implementation of handling `.ng` easy.
4. I don't have any strong feelings for the metadata function so I'm open to ideas surrounding them.

**Reply by joshuamorony** [reply]
Was the reason for moving away from the original `export const metadata` approach to more easily distinguish between component/pipe/directive?

Would it be realistic to do something like this as an alternative to defineDirectiveMetadata etc:

```ts
export const metadata = {
  type: 'directive',
  etc...
}
```

Perhaps for convenience the type could be component by default, so you could still just do this for components:

```ts
export const metadata = {
  selector: 'app-root'
}
```

**Reply by nartc** [reply]
Yes. Another reason is we can kinda provide type safety for a limited set of properties that out current implementation supports. 

We can go with a combination of:

```html
<script lang="ts" ng="Directive">
const metadata = {}
</script>
```

but again, no type info for `metadata` here. I'm reluctant to provide any kind of imports from analogjs at this point.

**Reply by joshuamorony** [reply]
I also agree with you about the pipes not really being necessary, everything going on here is already pretty opinionated/drastic - would it be going too far to just not implement pipes and instead provide a recommendation on using computed for pipes. Not having to account for pipes could simplify things here.

A pipe in this format doesn't feel quite right to me, but the component/directive concept feels natural imo, e.g we have a directive:

```html
<script lang="ts">
  stuff
</script>
```

and a component, which Angular developers are used to thinking of as a "directive with a template":

```html
<script lang="ts">
  stuff
</script>

<template>
  stuff
</template>
```

I am going to read through your PR properly, so my apologies if I am just being ignorant of something here, but I assume we could then just treat the presence/non-presence of a `<template>` to distinguish between a directive and a component.

Since the `@Component` metadata is extended from `@Directive` there is much more overlap in the typing, perhaps it would then be suitable to just use the same metadata approach for both with the component typing, e.g something like:

**highlight.ng**
```html
<script lang="ts">
  metadata({
    selector: 'p[highlight]'
  })
</script>
```

**some-component.ng**
```html
<script lang="ts">
  metadata({
    selector: 'app-some-component'
  })
</script>

<template></template>
```

I suppose the downside here would be that the typing would allow supplying component only properties to directives.

**Reply by nartc** [reply]
Relying on the existence of `<template>` to determine whether it's a Component or Directive sounds nice. I'd also lean towards getting rid of Pipes. Thoughts @brandonroberts ?

**Reply by brandonroberts** [reply]
I like both ideas. If using a computed is going to be preferred vs pipes, we should direct people to those.

**Reply by brandonroberts** [reply]
I also prefer `defineMetadata` vs `metadata` as its too concise IMO,  but not strongly.

**Reply by ajitzero** [reply]
> getting rid of Pipes

This is very reasonable for almost every pipe, but I'm unsure how to approach the async pipe? `signal.set(..)` inside a manual subscribe()? What am I overlooking?

Edit: nvm, we can probably use [toSignal](https://angular.io/api/core/rxjs-interop/toSignal).

**Reply by nartc** [reply]
@ajitzero you're overlooking toSignal

**Reply by nartc** [reply]
Feedbacks are implemented here: https://github.com/analogjs/analog/pull/826

**Comment by tayambamwanza** [comment]
Just here to bikeshed that I'm pro ```<template>``` tag 👀 but otherwise I'll definitely be experimenting with this.

**Reply by brandonroberts** [reply]
We landed on using a `<template>` because it's more straightforward to parse, but I can see the value in either approach. It also helps separate a component from a directive more clearly.

**Reply by lukaszpiotrluczak** [reply]
I could agree with reasoning. I only wonder if this is really a problem to parse when we don't have `<template>` tag?

If parser have default context of template then when `<script>`, `<style>` or any other support tag is used it will just change context to properly parse that part and then go back.

It will even allow us to add multiple `<script>'s` in component, or maybe even give them a scope and use it to inline event handler.

**Reply by nartc** [reply]
It's not impossible. We can always iterate on the syntax at this moment. Feel free to include any kind of snippets that resonate with your ideas and we can discuss further.

**Reply by Jordan-Hall** [reply]
```
<script lang="ts">
  import {
    DestroyRef,
    inject,
    input,
    EventEmitter,
    effect,
    ViewChild,
    ElementRef,
    afterNextRender,
  } from '@angular/core';

  import { myFunc } from './export-stuff.ng';

  defineMetadata({
    queries: {
      divElement: new ViewChild('divElement'),
    },
    exposes: [myFunc],
  });

  let divElement: ElementRef<HTMLDivElement>;

  afterNextRender(() => {
    console.log('the Div', divElement);
  });

  const { foo: aliasFoo = 'tran', ...rest } = { foo: 'chau' };
  const [a, b, , c = 5, ...restArray] = [1, 2, 3];

  const text = input('');

  effect(() => {
    console.log('text changed', text());
  });

  const clicked = new EventEmitter<MouseEvent>();

  inject(DestroyRef).onDestroy(() => {
    console.log('hello destroyed');
  });
</script>

  <h1>Hello.ng again</h1>
  <p>{{ a }}</p>
  <p>{{ b }}</p>
  <p>{{ c }}</p>
  <p>{{ aliasFoo }}</p>
  <p>Text from input: {{ text() }}</p>
  <button (click)="clicked.emit($event)">Emit The Click</button>
  <div #divElement>my div</div>
  <p>From imported function: {{ myFunc() }}</p>


<style>
  :host {
    display: block;
    padding: 1rem;
    border: 1px dashed red;
    border-radius: 0.5rem;
  }
</style>

```

I was looking at it for langauge service. If you take the start and end loc for style and script. Anything else is the HTML

**Comment by joshvanallen** [comment]
Just a quick flyby question...is specifying the script language truly needed? Is that a requirement for Vite or something else?

**Reply by brandonroberts** [reply]
Its required as it provides better Intellisense within an IDE

**Reply by Jordan-Hall** [reply]
@brandonroberts  what IDE are you using/plugin. When im using it with TS im getting no extra help it doesnt understand typings :/

**Reply by nartc** [reply]
@Jordan-Hall VSCode with Volar (Vue Language Service (Volar)) extension and set the `.ng` files to `.vue` files.
<img width="650" alt="image" src="https://github.com/analogjs/analog/assets/25516557/79872ec5-0459-4e96-ac48-82439b1fb22a">
<img width="222" alt="image" src="https://github.com/analogjs/analog/assets/25516557/65c599af-051f-45ff-838b-3eb6e223637d">

**Reply by Jordan-Hall** [reply]
Thank you I assumed it would be like that. Is anyone working on a actual angular volar language service to support .ng

**Comment by goetzrobin** [comment]
I love this! This is incredible work! 

This has me dreaming of supporting MDX rendering with something like this:

```html
<script>
import Hello from './hello.ng';
</script>

<markdown>
  # Hello World
 <Hello />
</markdown>
```

Do you guys think this would be possible?

**Reply by joshvanallen** [reply]
Might be a bit of an overengineering aspect for the time being but i wonder if you can specify the document root element. `template` being just html and `markdown` as a special transformer

**Reply by brandonroberts** [reply]
Definitely have considered this also, but want to get the foundation down first. I floated this idea also.

```html
<template lang="md">
</template>
```

**Reply by lukaszpiotrluczak** [reply]
Hmm... this situation defends idea to use `<template>` approach.

Will it require much work to add possibility to easily extend with other formats than html, md?

**Reply by nartc** [reply]
Another option is a different file extension that follows Markdown syntax rather than HTML (so no script tags and co.)

**Reply by goetzrobin** [reply]
I do see the value of seeing what type of content is in the file based on the file ending...

But I also love how good this looks:

```html
<script lang="ts">
import Hello from './hello.ng';
</script>

<template lang="md">
  # Hello World
 <Hello />
</template>
```

But no matter what the second this is supported I'll be all over it! Happy to help make it a reality, too, if there is any way I can help contribute!

**Reply by Jordan-Hall** [reply]
I like to add saying ng is a analog framework/authoring. I think using lang on all three tags would be useful and beneficial.  The example  by @goetzrobin seems like a good way to go about this and keeps learning curve small. Better than mutiple file types where it can get people confused.

**Comment by luishcastroc** [comment]
I love it, I'm pro <template> and I would love to help on this if needed, one question do we need to have a .ng for directives and pipes as well? That looks a little odd but other than that this looks fantastic

**Reply by brandonroberts** [reply]
We've already scoped it down to just components and directives, and will remove `.ng` pipes.

**Comment by ducin** [comment]
What are the potential possibilities of making it a pre-processor so that standard Angular apps can make use of it as well? Have you identified any critical obstacles, other than lack of supporting features such as deferrable views, inputs, outputs etc?

**Reply by brandonroberts** [reply]
I'll say upfront that this won't be possible with standard Angular apps because there is no direct access to the underlying Angular compilation process and this is by design in Angular. Its not a limitation of Angular, but an agreed-upon constraint by the Angular team. We are not sidestepping that process either, as we're still going through the Angular compilation process in the end.

These will only be available in Analog for the foreseeable future, unless Angular introduces a different authoring format, which they said they are not doing at the moment.

There aren't any critical obstacles at the moment, but additional features like signal inputs will be added when they are added to Angular upstream.

**Comment by Yberion** [comment]
Hello, it is going to be only available as SFC?

**Reply by brandonroberts** [reply]
Unless there is some compelling reason otherwise, yes. If you want to split up your component into separate files, you can still use a class component.

**Comment by nartc** [comment]
I thought I saw a notification about Generics Component. That is definitely something I had on my mind but I didn't have a concrete idea as to how to implement it. So I'm open-mind about it

**Reply by brandonroberts** [reply]
I saw it also, maybe I misunderstood the context of the question.

**Comment by krivanek06** [comment]
I like it very much. For lifecycle hooks I think we would need some eslint support. You can accidentally make a mistake my writing `inInit()` instead of `onInit()` ... maybe somehow incorporate not necessarily implementing interfaces, but somehow use them. For example the `ControlValueAccessor` is quite a common interface to implement

**Reply by nartc** [reply]
Thank you for the feedback.

1. A LSP support might be better than a linter. That is something we have in mind, but will see if the progress of `.ng` warrants the time put into creating a LSP
2. Good point regarding `ControlValueAccessor`. We'll need to figure out something. Feel free to shoot us ideas if you come up with something too.

**Reply by Jordan-Hall** [reply]
If you need to implement something wouldn't it be better to use class for it without having to define the metadata? Alternatively you could do something like this when lang=ts

<script lang="ts" extends="" implement="">

I think doing it on the script tag better than in defineMetadata for example. Plus you can easily abstract it from the tags.

**Reply by brandonroberts** [reply]
Yea, I don't think we should be extending the script tag that way. Seems too far off from its intended usage. If you need to do class-based things, use a class component. Let's not build a pseudo-class language in the script tag itself.

**Reply by Jordan-Hall** [reply]
I agree with, i dont like the idea of building into script tag. I guess the only question is around export. I think personally dont use .ng file for anything that needs to extend or implement.

**Reply by nartc** [reply]
re:ControlValueAccessor https://ngxtension.netlify.app/utilities/forms/control-value-accessor/ someone actually submits a utility to implement `ControlValueAccessor` without `implements ControlValueAccessor`

**Comment by KevTale** [comment]
That looks promising! GG guys 👏

Here's a couple of questions:
1. Are all variables in `<script>` exported in `<template>`? How can I prevent the template to access some variables?
2. How does routing works for `.ng` files as it was previsously handled by the class being exported by default? Can we expect maybe something like `some.route.ng` and `some.component.ng`? I like the fact that you can visually determine what is the purpose of a file just by looking at its name.
3. Can I place the `<script>` tag in the bottom of the file rather than at the very top?

**Reply by nartc** [reply]
1. Currently, there is no syntax available to determine which variables can be accessed on the template. While we technically have the option to use `export const ...` to only expose certain variables to the template, it may not be worthwhile.
2. I'll let @brandonroberts answer this
3. Yes, you can place the `script` tag at the bottom.

**Reply by brandonroberts** [reply]
1. Agreed. I would reserve `export` for inputs/outputs, but we may be able to target those based on the function being used `input()` or `output()` in the future. We'd have to come up some conventions to keep variables out of the template scope.
2. We could support the same naming convention as `.page.ng` instead of `page.ts` to support the file-based routing. We do use some conventions based on the file names for "auto-imports" though.

**Comment by louiss0** [comment]
I'd like to be able to have multiple templates in one file. 
Can that be a thing. Svelte has it.

**Reply by nartc** [reply]
Can you elaborate on the use-case?

**Reply by brandonroberts** [reply]
Svelte doesn't have a notion of a `<template>` tag as far as I know, so how can there be multiple?

**Reply by louiss0** [reply]
I'm talking about ng-template and ng container. Without having to import anything.

```vue

<template>
<ng-container *ngTemplateOutlet="greet"></ng-container>
</template>

<ng-template #greet>
Hello World  
</ng-template>
```

**Reply by nartc** [reply]
I can make NgTemplateOutlet auto-imported. For this case, you'd still want the ng-template to be inside of the template tag

**Reply by louiss0** [reply]
Please try to find a way to not make this the case the code would look extremely disgusting and it would be difficult to sort from the main template.

**Reply by nartc** [reply]
the <template> tag acts as a boundary only. Literally everything inside it is brought over to the "template" property in the component's metadata.

**Reply by louiss0** [reply]
Okay

**Comment by marcus-sa** [comment]
Shouldn't `inject(DestroyRef).onDestroy` just be `onDestroy` without having to import or inject anything?

**Reply by nartc** [reply]
It is now. But I guess Brandon would have to cut another release to include that change.

**Reply by brandonroberts** [reply]
The latest beta is published `0.2.30-beta.3` and includes the updates

**Comment by vneogi199** [comment]
Not sure if I missed it, but I would to see variables/enums directly accessible in the template without creating an instance of them in the script section.

```svelte
<script lang="ts">
  import {myVar, ColorEnum} from 'another-one.ng';
</script>


<template>
Variable is: {{myVar}}

Is variable blue: {{myVar === ColorEnum.BLUE}}
</template>
```

**Reply by nartc** [reply]
Currently, we don't analyze the template to check thr bindings so for things like imported variables (lile an emum) or global symbols (like Math), you'd want to use `exposes` to expose them on the template (again, @brandonroberts might need to cut a new version 😅)

```html
<script lang="ts">
defineMetadata({
  exposes: [Math, MyEnum]
})
</script>
```

**Reply by brandonroberts** [reply]
The latest beta is published `0.2.30-beta.3` and includes the updates

**Reply by vneogi199** [reply]
I dont know if its technically possible to do something like:

`templateContext = {...currentTemplateContext, windowGlobals, importsFromScriptTag, exportedVariablesFromScriptTag}`

this way these will be available within template context.

**Reply by nartc** [reply]
It is technically possible to do anything that is within the JS syntax 😅

**Comment by xocomil** [comment]
Wow! I unplug for a few weeks and you guys drop something this huge! This looks amazing and I am excited to see where your ideas lead. I'm curious about how feasible functional components become with the helper functions. For example, would the following be possible?

```ts
  const component1 = createComponent(withMetadata({imports: [ signalStore, myPipe]}), withTemplate(`<h1>{{myStore.value()}}</h1>`);
  const component2 = createComponent(...);
```

There is some "magic" I'm glossing over here, but I'm more curious if you think the idea is feasible given the research you have done with the compiler.

I love this idea. I think it will make Angular more accessible.

**Reply by nartc** [reply]
If you are able to provide a full example, it'd help a bit more. Also, does this have anything to do with `.ng` files or just a suggestion at yet a different author format?

**Reply by xocomil** [reply]
Mainly curiosity on my part. I'm experimenting with some ideas right now. I was hoping to steal some of yours... 🤣

**Comment by marcus-sa** [comment]
I don't know if it's useful to you, but you can actually patch the Angular compiler to support custom transformers.
I've done something similar that you can see [here](https://github.com/deepular/deepular/tree/face9dd4c9222f7365b532d4942a3d5af1fe2506/packages/compiler) and [here](https://github.com/deepular/deepular/tree/main/packages/compiler/src/patches).

**Reply by brandonroberts** [reply]
Thanks. We're not trying to maintain a fork of the Angular compiler though

**Comment by joshvanallen** [comment]
keep toying away with this and thinking about a few things:
- for the logic that's outside of a lifecycle function, where is that code placed? Is it in the constructor or placed on the class itself (I'm assuming at the end of the day this gets translated into a class)?
- NgModule support? Granted we are two major versions since standalone was properly released, there's still libs out there that rely on it. more of a clarifying question/setting expectations for everyone

**Reply by nartc** [reply]
1. Top-level expressions (i.e: `effect(() => {})`) are placed in the "class" constructor (you can take a look at the test snapshot for `ng.ts`)
2. `NgModule` would still work. 
   ```html
   <script lang="ts">
   defineMetadata({
     imports: [CommonModule]
   })
   </script>
   ```

**Comment by MitchSmoot** [comment]
I remember seeing Brandon's politely worded tweet earlier about lack of functional components "pushing angular into niche markets". My initial reaction was: why change it if it isn't broken? But after seeing this implementation, even after sharing some initial concerns in a response to Joshua's tweet, I'm really warming up to this. Plus, we get to flex on the tech blogosphere which is always a good thing.

I think it is a fantastic addition for angular devs who know what they are doing. But I have not gotten to the point of thinking it should be recommended to new developers. Having multiple, interchangeable, ways of doing a thing makes it harder to learn. All of the tutorials, examples, docs are in the original syntax, and I don't think it would be an enjoyable experience for a new dev who doesn't understand the workings and flow of the framework to have to try and write in this new syntax when all their online resources are written the other way. BUT if we can create really good documentation that include more substantive examples, I think we can alleviate that concern. I would be happy to help contribute to such documentation. 

I see myself using both authoring forms initially, old way for pages, new way for components on pages, or small components. Could switch wholly to the new format if it feels nice in practice.
**The use cases I am most looking forward to**:
- three.js or angular-three components
- Astro x Analog applications

This authoring syntax seems like a home run for Astro+Analog projects. I don't think I need to say more.
Components like those built with angular-three that already prefer single-file format are a natural fit, and I cant wait to use them in my current project. Will there need to be a special angular-three integration or will it work with this out of the box? @nartc I would love to hear your perspective on this. I'm almost at the point where I would donate my kidney to try this out with angular-three 2.0, haha.

**Reply by nartc** [reply]
`angular-three` should work out of the box for this. We might be able to do something extra to `.ng` file if that file needs `CUSTOM_ELEMENTS_SCHEMA` or something but they should just work out of the box. 

Now that you remind me, I want to build `angular-three` with the new format as well but building the library is different 😅.

**Comment by ajitzero** [comment]
Can we skip defining the selector for components?

It would still be needed for directives, but components are already being renamed with `import Hello from './hello.ng';`  as `<Hello />` instead of `<app-hello />`. If needed internally, we could use the file name (maybe?) to generate the component name.

I would love to be able to skip the wrapper element à la React Fragments, but that might be beyond the scope here.

Thoughts?

**Reply by nartc** [reply]
Yes, by default the file name will be used as the selector

```
hello.ng -> selector: 'hello,HELLO,Hello',
another-one.ng -> selector: 'another-one,AnotherOne,ANOTHER_ONE'
```

you can omit selector for components.

**Reply by nartc** [reply]
re:Fragment
Yes, this is beyond the scope. We don't do anything different in terms of rendering. That will have to come from Angular upstream

**Comment by Jordan-Hall** [comment]
Currently the styles are added to the template in the meta. Should the styles be moved to the styles in the meta to support multiple <style> tags or is it only 1 thats allowed?

**Reply by brandonroberts** [reply]
We tried putting them the `styles` instead of combining them in the `template` but they weren't being applied correctly. What's the use case to need more than one?

**Reply by Jordan-Hall** [reply]
My thinking was because Angular allows multiple. Do you mind if I take a look at the styles and combining, Looking through the code means you wouldn't be able to do SCSS for example right? Unless processing it inside of analog. If it going to be process inside of analog wouldn't you need a lang attribute to styles?

**Reply by Jordan-Hall** [reply]
![image](https://github.com/analogjs/analog/assets/2092344/840926e9-4fff-445e-b653-8f8afc35bd5c)
Another reason for asking is to block it from being allowed

**Comment by Jordan-Hall** [comment]
Proposal drop then .ng and rename it. This allows angular to use the file name in future without and clashes especially if this comes out of experimental for analog

**Reply by brandonroberts** [reply]
Good idea. We were already considering this prior to yesterday

**Reply by Jordan-Hall** [reply]
I thought you would. Could be cheeky and do something like .ngx but would love to see this more oauh as analog with angular di and rendering system. This changing file ext be helpful and free up for angular in future
